std/file.d

```diff
@@ -465,30 +467,98 @@ version (linux) @safe unittest
     assert(read(deleteme) == "abcd");
 }
 
-/********************************************
-Read and validates (using $(REF validate, std,utf)) a text file. $(D S)
-can be a type of array of characters of any width and constancy. No
-width conversion is performed; if the width of the characters in file
-$(D name) is different from the width of elements of $(D S),
-validation will fail.
-
-Params:
-    name = string or range of characters representing the file _name
+/++
+    Read and validates (using $(REF validate, std, utf)) a text file. S can
+    be a type of array of characters of any width and constancy. No width
+    conversion is performed; if the width of the characters in file name
+    is different from the width of elements of S, validation will fail.
+
+    Before validating the characters, the file is first checked for a
+    $(HTTP https://en.wikipedia.org/wiki/Byte_order_mark, BOM). If it is present
+    and matches the encoding of the requested string type, then it is stripped
+    from the output, but if it doesn't match, then a
+    $(REF BOMException, std, encoding) is thrown.
+    $(REF BOMException.bomSeq, std, encoding) will then indicate what the actual
+    encoding was, and the program can then request the string type that matches
+    that encoding.
 
-Returns: Array of characters read.
+    Params:
+        name = string or range of characters representing the file _name
 
-Throws: $(D FileException) on file error, $(D UTFException) on UTF
-decoding error.
- */
+    Returns: Array of characters read.
 
-S readText(S = string, R)(R name)
-if (isSomeString!S &&
-    (isInputRange!R && !isInfinite!R && isSomeChar!(ElementEncodingType!R) || isSomeString!R) &&
-    !isConvertibleToString!R)
+    Throws: $(LREF FileException) if there is an error reading the file,
+            $(REF BOMException, std, encoding) if there is a BOM at the front
+            of the input, and it does not match the requested encoding, and
+            $(REF UTFException, std, utf) if the text contains invalid Unicode.
++/
+S readText(S = string, R)(auto ref R name)
+if (isSomeString!S && (isInputRange!R && !isInfinite!R && isSomeChar!(ElementType!R) || is(StringTypeOf!R)))
 {
-    import std.utf : validate;
+    import std.algorithm.searching : startsWith;
+    import std.encoding : BOM, BOMException;
+    import std.exception : enforce;
+    import std.format : format;
+    import std.utf : UTFException, validate;
+
+    static if(is(StringTypeOf!R))
+        StringTypeOf!R filename = name;
+    else
+        auto filename = name;
+
+    immutable bomSeq = readBOM(filename);
```

Comment:
Won't opening the file twice break reading from streams? Better read the file first, then use `getBOM`. `readBOM` seems unnecessary.

Reply:
I forgot about streams. I did that with the idea that it's quite inefficient to read in the entire file just to then throw an exception due to the fact that the actual encoding doesn't match the requested encoding. But I guess that if someone is really worried about that, they can call `readBOM` themselves.

Reply:
I think we shouldn't add `readBOM` 1) because we're not going to use it ourselves anyway 2) because it doesn't look like good practice due to being susceptible to race conditions.